CMake/FindMKL.cmake

```diff
@@ -2,42 +2,106 @@
 # This needs a lot of work to make it robust
 INCLUDE( CheckCXXSourceCompiles )
 
+if ( NOT CMAKE_CXX_COMPILER_ID MATCHES "Intel" )
+  # if not intel and MKL_ROOT not set
+  if (NOT MKL_ROOT)
+    find_path(MKL_ROOT "mkl.h"
+      HINTS ${MKLROOT} ${MKL_HOME} $ENV{MKLROOT} $ENV{MKL_ROOT} $ENV{MKL_HOME}
+      PATH_SUFFIXES include)
+    string(REPLACE "/include" "" MKL_ROOT ${MKL_ROOT})
+  endif (NOT MKL_ROOT)
+  if (NOT MKL_ROOT)
+    message (FATAL_ERROR "ENABLE_OUTSIDE_MKL is TRUE and mkl not found. Set MKL_ROOT." )
+  endif (NOT MKL_ROOT)
+
+  #Do we support ilp?
+  set(MKL_FIND_LIB "libmkl_intel_lp64${CMAKE_SHARED_LIBRARY_SUFFIX}")
+  set(SUFFIXES lib lib/intel64)
+  find_path(MKL_LINK_DIRECTORIES name "${MKL_FIND_LIB}" HINTS ${MKL_ROOT}
+    PATH_SUFFIXES ${SUFFIXES})
+  if (MKL_LINK_DIRECTORIES-NOTFOUND)
+    message(FATAL_ERROR "MKL_LINK_DIRECTORIES not set. ${MKL_FIND_LIB} "
+      "not found in MKL_ROOT/(${SUFFIXES})")
+  endif (MKL_LINK_DIRECTORIES-NOTFOUND)
+  message("MKL_LINK_DIRECTORIES: ${MKL_LINK_DIRECTORIES}")
+  set(SUFFIXES include)
+  find_path(MKL_INCLUDE_DIRECTORIES name "mkl.h" HINTS ${MKL_ROOT}
+    PATH_SUFFIXES ${SUFFIXES})
+  if (MKL_INCLUDE_DIRECTORIES-NOTFOUND)
+    message(FATAL_ERROR "MKL_INCLUDE_DIRECTORIES not set. \"mkl.h\" not found in MKL_ROOT/(${SUFFIXES})")
+  endif (MKL_INCLUDE_DIRECTORIES-NOTFOUND)
+  message("MKL_INCLUDE_DIRECTORIES: ${MKL_INCLUDE_DIRECTORIES}")
+  set(MKL_FLAGS "-m64")
+  set(MKL_LINKER_FLAGS "-L${MKL_LINK_DIRECTORIES} -Wl,-rpath,${MKL_LINK_DIRECTORIES}")
+  set(MKL_FLAGS "-I${MKL_INCLUDE_DIRECTORIES}")
+  set(MKL_LIBRARIES "-lmkl_intel_lp64 -lmkl_sequential -lmkl_core -lpthread -lm -ldl")
```

Comment:
We need threaded MKL.

Reply:
really?  Generally openmp codes run into performance problems is they also let mkl thread it self.  This just means there is not threading within mkl itself.

Reply:
Trust us.
Remember -mkl is default -mkl=parallel
https://software.intel.com/en-us/mkl-linux-developer-guide-using-the-mkl-compiler-option

Reply:
Ok, I trust you for the intel build, but it requires (as far as I can see) intels openmp and I haven't tested that at all on clang or gcc. I'm not using intels openmp or mixing openmp libraries anywhere else. I'd rather file that as an issue with this PR and deal with it as a separate concern.

Reply:
I'm also skeptical of linking with the threaded MKL.    It seems safer and more general to link with the sequential version.

Reply:
Only the matrix inversion used by the optimizer and FFT needs threaded MKL. All the rest MKL calls are inside OMP region and MKL will not turn on nested threading by default. I personally don't have any concern but if you guys prefer sequential. Just do sequential at the moment.